lasagne/layers/conv.py

```diff
@@ -249,7 +249,7 @@ def get_output_for(self, input, input_shape=None, **kwargs):
                                       filter_shape=self.get_W_shape(),
                                       border_mode='full')
             shift = (self.filter_size[0] - 1) // 2
-            conved = conved[:, :, shift:input.shape[2] + shift]
+            conved = conved[:, :, shift:-shift]
```

Comment:
> I'm not sure how to even define a 'same' convolution for even kernels...

We've stated in the documentation how we define it:
```
'same' pads with half the filter size on both sides (one less on the second side for an even filter size). When stride=1, this results in an output size equal to the input size.
```

> it's not covered by the tests, and the current implementations fail also.

Oh yes, that's something were our coverage reporting fails, because it only checks whether each line is touched. We should test it, and have convolution implementations that don't support it raise a `NotImplementedError`.

In Theano's GPUCorrMM, I've added a padding mode `"half"` that always pads with half the filter size.

> I think this could work, I'll try profiling both ways later.

Alternatively, you could have a `left = (self.filter_size[0] // 2) or None` and `right = -(self.filter_size[1] - left) or None`, then do `conved = conved[:, :, left:right, :]`, something like that? I think I planned to do that in #374, but then I remember it turned out to be more complicated than anticipated -- or maybe that was for the Conv2DDNNLayer et al?